src/content/post/2024-04-05-rdf-shacl-and-kuzu.md

```diff
@@ -0,0 +1,343 @@
+        Result Path: kz:name
+        Message: String length not >= Literal("5", datatype=xsd:integer)
+```
+
+The violation can easily be fixed by setting the minimum length of the `name` property to a smaller value, such as 2.
+When this is done, the validation passes.
+
+```turtle
+Validation Report
+Conforms: True
+```
+
+## Visualize the RDF graph in Kùzu Explorer
+
+When building and validating RDF graphs, the ability to get visual feedback is quite useful. Kùzu
+Explorer is a web-based interface that allows you to visualize RDF graphs and query them using Cypher (no knowledge of SPARQL required).
+The instructions to launch Kùzu Explorer and connect to an existing database are shown in [the docs](https://docs.kuzudb.com/visualization/).
+
+```bash
+docker run -p 8000:8000 \
+    -v /Users/prrao/code/kuzu-rdflib/db:/database \
+    --rm kuzudb/explorer:latest
+```
+
+In a nutshell, Kùzu's RDFGraphs extension creates four distinct tables when the RDF data is loaded into the database:
+
+- `UniKG_l`: A node table that contains literals
+- `UniKG_r`: A node table that contains resources, where the primary key is the unique IRI
+- `UniKG_lt`: A relationship table that contains resource-to-literal triples
+- `UniKG_rt`: A relationship table that contains resource-to-resource triples
+
+The following Cypher query can be run to visualize the entire graph:
+
+```cypher
+MATCH (s)-[p]->(o)
+RETURN *;
+```
+
+![](/img/rdf-shacl-kuzu/demo-rdf-viz.png)
+
+As can be seen, the graph structure is identical to that shown earlier, in the conceptual representation.
+
+The yellow edges represent resource-to-literal relationships (`UniKG_lt`), while the red edges represent
+resource-to-resource relationships (`UniKG_rt`). We can inspect the schema of the RDF graph, including
+each table's primary keys, visually, by clicking on the "Schema" tab in Kùzu Explorer.
+
+![](/img/rdf-shacl-kuzu/demo-rdf-schema.png)
+
+## Querying the RDF graph with Cypher
+
+Further Cypher queries can be run on the RDF graph, that perform the same operations as their SPARQL equivalents.
+In the example below, we want to run a query to only return students named "Karissa".
+
+```cypher
+// Run using Kùzu Explorer
+WITH "http://kuzu.io/rdf-ex#" as kz
+MATCH (s {iri: kz + "Karissa"})-[p1 {iri: kz + "name"}]->(l)
+WHERE (s)-[p2]->(o {iri: kz + "student"})
+RETURN DISTINCT s.iri, p1.iri, l.val;
+```
+
+The above query is functionally equivalent to this SPARQL query that can be run in RDFLib:
+
+```sparql
+# Run using RDFLib
+PREFIX kz: <http://kuzu.io/rdf-ex#>
+SELECT DISTINCT ?src ?name
+WHERE {
+  ?src a kz:student .
+  ?src kz:name ?name .
+```
+
+Both queries would return the following result:
+
+```
+                        s.iri                       p1.iri  RDF_VARIANT
+http://kuzu.io/rdf-ex#Karissa   http://kuzu.io/rdf-ex#name      Karissa
+```
+
+As can be seen, you can choose the most appropriate query language to analyze your data, depending on your
```

Comment:
I would add a note somewhere saying that you can extend Kuzu's RDFGraphs with other property graph, and use Cypher to query both your triples and other property graphs uniformly with Cypher. And point them to this documentation page: https://docs.kuzudb.com/rdf-graphs/rdfgraphs-overview/#querying-of-regular-node-and-relationship-tables-and-rdfgraphs.